File: 153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.cpp
#pragma GCC target("tune=native")
#pragma GCC optimize("Ofast")
//#pragma GCC optimize("Os")
// Code here is optimized for size

static const auto fastIO = []() {
    std::ios::sync_with_stdio(false);
    std::cin.tie(nullptr);
    std::cout.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int findMin(vector<int>& arr) {
        int low = 0, high = arr.size() - 1;
        int ans = INT_MAX;

        while (low <= high) {

            int mid = (low + high) / 2;

            // Sorted Portion ka min utha lo 
            if (arr[low] <= arr[mid]) {
                ans = min(ans, arr[low]);
                low = mid + 1;
            } 
            else {
                // Agar right side sorted hai toh mid uska min hoga
                ans = min(ans, arr[mid]);
                high = mid - 1;
            }
        }
        return ans;
    }
};

File: 153-find-minimum-in-rotated-sorted-array/find-minimum-in-rotated-sorted-array.java
class Solution {
    public int findMin(int[] nums) {
        int l = 0, r = nums.length - 1;
        int ans  = Integer.MAX_VALUE;

        while(l <= r){
            int mid = l + ( r- l) / 2;

            if(nums[l] <= nums[mid]){
                ans = Math.min(nums[l], ans);
                l = mid + 1;
            }
            else{
                ans =Math.min(ans, nums[mid]);
                r = mid - 1;
            }
        }

        return ans;
    }
}